[WIP] Unlock Membership Tiers for Kickback

edsonayllon opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Issue number
<!--- If there is an associated github issues, please specify here -->

#209 

## Description
<!--- Describe your changes in detail -->

> @julien51  The Kickback team wants to provide paid benefits for people willing to deploy their own kickback contract. In order to do that, the kickback will deploy one or more locks (one for each level of benefits).
>
>The front-end form will then ensure that the user owns a key to any of the kickback locks in order to deploy their kickback contract. If the user does not have a key, they are offered the ability to purchase a key to any of the lock (either thru crypto or credit card) thru Unlock.
> [...]

See issue #209 linked.

I'm this PR, unlock displays page, as seen in the image below, requesting membership to view requested content. Once verified, form to deploy kickback contract is displayed to the user with membership status.

## List of features added/changed
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- Adds membership tiers

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->

Currently undergoing testing. Work in progress.

## Screenshots (if appropriate):

<img width="1293" alt="Screen Shot 2019-12-09 at 9 06 30 PM" src="https://user-images.githubusercontent.com/34966228/70488802-d3a76b00-1ac7-11ea-8afe-c2e9bf923e07.png">

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My code implements all the required features.

makoto commented:
Hi, @edsonayllon . Thank you for the submission. When I tested locally, Clicking the "Unlock membership? didn't do anything. 

I can also see "An error occurred on client" message (highlighted in red).

<img width="740" alt="Screenshot 2019-12-10 at 21 25 29" src="https://user-images.githubusercontent.com/2630/70570457-d8d1ec00-1b93-11ea-96ea-8c30e18a30f0.png">

I can also see some error messages on console when I first load the /create page cc @julien51 

<img width="582" alt="Screenshot 2019-12-10 at 21 29 55" src="https://user-images.githubusercontent.com/2630/70570714-51d14380-1b94-11ea-8971-7f82c9061349.png">

edsonayllon commented:
@makoto Yes, this PR isn't ready for review yet (marked as Work in Progress). 

I ran into this issue yesterday. I talked to @julien51 and it seems Unlock's paywall app is only deployed on Rinkeby testnet, however, the addresses provided were for Kovan. Julien's provided me keys for Rinkeby which I'll test soon. 

The Readme for the app includes instructions for Kovan, but `package.json` shows Rinkeby is available. Is there any documentation available for using Kickback on Rinkeby?

edsonayllon commented:
Sorry, closed the PR by accident. 

In 292b7b0 paywall now is functional when using Rinkeby. 

<img width="1368" alt="Screen Shot 2019-12-10 at 10 14 25 PM" src="https://user-images.githubusercontent.com/34966228/70593141-ed20e380-1ba9-11ea-8e8a-05e08d2ffa1f.png">

A remaining task having paywall load for the correct network, either Rinkeby or Mainnet. 

@julien51 stated `src/config/env.json` can somehow be used in #209.

edsonayllon commented:
Also, for the checks, it looks like the build is successful, but fails from lacking credentials at deployment. 

<img width="768" alt="Screen Shot 2019-12-11 at 12 21 50 AM" src="https://user-images.githubusercontent.com/34966228/70593805-64577700-1bac-11ea-8411-64ed8d181e81.png">

edsonayllon commented:
682d830 Network for Unlock is selected based on `NODE_ENV` which can be `development`, `staging`, or `deployment` depending on build or start scripts. 

Dynamically selecting the network for Unlock was a bit tricky. Unlock is configured in `public/index.html`, as seen through the docs and their React example on the Unlock Github. I didn't see a way to access `env.json` from the public folder. Discovering the current network by importing the [`dist`](https://github.com/ethereum/web3.js/tree/1.x/dist) of web3.js would increase load times of the app, so I avoided that as well. 

To work around these, seeing the testnet version of Unlock would only be accessed during development or staging, and the mainnet verion of Unlock would be accessed in deployment, I set dynamically choosing the network based on the [Environment Variable `NODE_ENV`](https://create-react-app.dev/docs/adding-custom-environment-variables/).

Again, currently Unlock Paywall is only compatible with Rinkeby and Mainnet, so testing should be done on Rinkeby. I've left the unlock keys for Mainnet empty, so make sure to enter those when your team creates them.

julien51 commented on public/index.html:
```diff
@@ -34,6 +34,70 @@
+    // unlock config must use var
+    var unlockProtocolConfig = {
+      locks,
+      icon: 'https://unlock-protocol.com/static/images/svg/unlock-word-mark.svg',
```
This could be changed to a Kickback logo ;)

edsonayllon replied:
Good idea. Changed in a60f5f7. It looks like only public urls are compatible with Unlock, so I used https://kickback.events/card.png.

makoto commented:
Hi, 

Can you write an instruction on how to test on my local machine? I just did `NODE_ENV=development yarn start` and opened my browser pointing network to Rinkeby but it shows "Wrong network", and pointing to mainnet does not seem to load locks.

edsonayllon commented:
@makoto `NODE_ENV` is defined automatically by `create-react-app`, seeing if either `react-scripts start` or `react-scripts build` is used. Starting with `yarn start` should have the environment default to `"development"`. Anything running `yarn build` should change the environment to `"production"`. 

The red message box indicating wrong network is configured by `src/config/env.json`. By default that's set to Kovan, and displays this error if Rinkeby is used. However, the Paywall only works in Rinkeby or Mainnet, as stated before. So switching to Rinkeby should load locks and allow purchasing, but will display "wrong network" on the Kickback app. Once a token is purchased in Rinkeby, switching to Kovan should still show the form, and remove the error message if configured to Kovan. Because Unlock is set to default to Rinkeby unless the app is built for production, which sets it to Mainnet, if the account owns the token in Rinkeby it will still display the field even if the account provider changes network.

edsonayllon commented:
*Note:* I changed the component `Create` to a functional component using hooks, as the Readme Contribution Style Guide said the app is transitioning in that direction anyway. 

I can change it back to a class component, if your team prefers.

makoto commented:
So is this the correct config?

```
$more src/config/env.json
{
  "API_URL": "https://kovan.api.kickback.events",
  "NUM_CONFIRMATIONS": 1,
  "ENV": "rinkeby"
}
```
I am still getting "Wrong network" on unlock.

<img width="560" alt="Screenshot 2019-12-12 at 09 33 58" src="https://user-images.githubusercontent.com/2630/70700391-91964900-1cc2-11ea-8e30-579361fe054b.png">

Also Is it possible to disable Unlock if `ENV=local` ? Otherwise we cannot test when we are testing against local server and ganache.

edsonayllon commented:
@makoto Can you verify that you pulled the latest commits? Also, what are the steps you are taking to run the app?

As for disabling Unlock, I can disable the paywall, and display the Create form by default if `ENV=local` in `env.json`. However, preventing the script from running in `public/index.html` would be more complicated. I'm guessing, however, you only need the form to show in ganache.

edsonayllon commented:
595316b Paywall is disabled if `env.json` sets `ENV` to local network, displaying "Create" form. 

Is something wrong with https://kovan.api.kickback.events/graphql? It's throwing errors in the console.

makoto commented:
Hi, @edsonayllon kovan was down but now up. You can also test with https://rinkeby.kickback.events/create . Still not fully up and running but at least you can test the account creation page.

makoto commented:
umm, I deployed your branch to both rinkeby and kovan but both are showing spinning message.

<img width="680" alt="Screenshot 2019-12-13 at 12 48 07" src="https://user-images.githubusercontent.com/2630/70801467-e57a5e00-1da6-11ea-971d-1dc514b05248.png">

```
{
  "DEPLOYER_CONTRACT_ADDRESS": "0x8fE91259DC1F723d01b87Cf29534d8484F70e7Bc",
  "API_URL": "https://rinkeby.api.kickback.events",
  "NUM_CONFIRMATIONS": 1,
  "ENV": "rinkeby",
  "GIT_COMMIT": "595316b9f82d516e64ffda03cffe7b5e0a82a9d1",
  "ROLLBAR_TOKEN": "e676d64e462b48d098a12db8a173598a",
  "BLOCKNATIVE_DAPPID": "27b3eac2-e46c-428a-9a0c-56cce2725d42",
  "MIXPANEL_ID": "28243587317e7b2d8a669dcce23302cb"
}
```

```
{
  "DEPLOYER_CONTRACT_ADDRESS": "0x8fE91259DC1F723d01b87Cf29534d8484F70e7Bc",
  "API_URL": "https://kovan.api.kickback.events",
  "NUM_CONFIRMATIONS": 1,
  "ENV": "kovan",
  "GIT_COMMIT": "595316b9f82d516e64ffda03cffe7b5e0a82a9d1",
  "ROLLBAR_TOKEN": "",
  "BLOCKNATIVE_DAPPID": "",
  "MIXPANEL_ID": "28243587317e7b2d8a669dcce23302cb"
}
```

Can you think of what would cause this?

makoto commented:
Actually I am seeing the same spinning on localhost if ENV=rinkeby (ENV=local will just render the event creation page as requested. Thanks!)

edsonayllon commented:
> @makoto umm, I deployed your branch to both rinkeby and kovan but both are showing spinning message. Can you think of what would cause this?

Yup. "deploy:rinkeby" or "deploy:rinkeby" actually configures unlock for mainnet. Currently, there are placeholders for mainnet keys, and not real addresses, which would cause an error in the configuration. When there's an error in the config, the state never updates from loading to locked or unlocked.

makoto commented:
Ah ok. so how do I test? I am a bit confused now as I cannot test on my machine nor on the deployed env. Would you mind how to set things up in our README so that anyone can just follow that?

edsonayllon commented:
> @makoto Ah ok. Would you mind adding the description of how to set things up in our README so that I can just follow that?

Added instructions in Readme. Deploy scripts now chooses appropriate Unlock network, limited to Rinkeby or Mainnet.  c1f1908.

makoto commented:
Hi.
I was able to test the purchase flow.
Once purchased and the create page is shown, would it be possible to display what lock has been purchased?

For the deployed environment, I again deployed to https://rinkeby.kickback.events/create but getting the spinning image when connecting to either rinkeby or mainnet.
Here is the config when deploying to rinkeby.
```
$yarn deploy:rinkeby
yarn run v1.15.2
$ yarn build:release:rinkeby && yarn now -f --local-config .deploy/now.rinkeby.json --public && yarn now alias --local-config .deploy/now.rinkeby.json && yarn now rm kickback-app-rinkeby --safe --yes
$ REACT_APP_ENV=rinkeby yarn setup --rinkeby && yarn build
$ scripts/setup.js --rinkeby
{
  "DEPLOYER_CONTRACT_ADDRESS": "0x8fE91259DC1F723d01b87Cf29534d8484F70e7Bc",
  "API_URL": "https://rinkeby.api.kickback.events",
  "NUM_CONFIRMATIONS": 1,
  "ENV": "rinkeby",
  "GIT_COMMIT": "c1f1908d7da54ccdb89da409eeb4dd0a1ea5b283",
  "ROLLBAR_TOKEN": "e676d64e462b48d098a12db8a173598a",
  "BLOCKNATIVE_DAPPID": "27b3eac2-e46c-428a-9a0c-56cce2725d42",
  "MIXPANEL_ID": "28243587317e7b2d8a669dcce23302cb"
}
```

Can you check  https://rinkeby.kickback.events/create  and see if you are seeing the same spinning image?

edsonayllon commented:
> @makoto  Once purchased and the create page is shown, would it be possible to display what lock has been purchased?

Something like this? 

<img width="1216" alt="Screen Shot 2019-12-15 at 3 02 39 PM" src="https://user-images.githubusercontent.com/34966228/70868557-ea2e4600-1f4f-11ea-9e51-55252650f6c8.png">

Added a1166a7.

> @makoto  For the deployed environment, I again deployed to https://rinkeby.kickback.events/create but getting the spinning image when connecting to either rinkeby or mainnet.

I tested and got the same thing. The environment variable was written to pass with `yarn setup` which should have instead been with `&& yarn build`. I corrected the placement, and added `cross-env` for compatibility with Windows. I added temporary keys for Mainnet, please replace these with your keys.

Fixed a1166a7.

jefflau commented on src/routes/CreateEvent.js:
```diff
@@ -13,42 +15,172 @@ const CreateContainer = styled('div')`
+
+  useEffect(() => {
+    // remove unlock in component unmount1
+    return () => {
```
Can this not be bundled with the other useEffect's return statement?

edsonayllon replied:
You're right. dbc56b9

jefflau commented on public/index.html:
```diff
@@ -52,13 +52,81 @@
+      Kovan lock is included for when kovan is supported by Unlock paywall
+      NOTE: Replace Mainnet keys with keys Kickback owns
+    */
+    const locks = {
```
Is there any way to abstract these into the main `src` directory. It's not very visible having the locks in the index.html.

edsonayllon replied:
I don't think so. I looked through Unlock's documentation, and looked at the React example in Unlock's Github. They had this config in `public/index.html`.  I also don't think there's a way to import a file from `src` to `public`, else I'd be using `env.json` instead of passing environment variables through `package.json`. Unless @julien51 has any suggestions.

edsonayllon replied:
Actually... I think this just adds a variable to `window`. Let me test something

edsonayllon replied:
Ok. Locks moved to `env.json`. Readme updated as well. 393de51

makoto commented:
Hi,I tried both local and https://rinkeby.kickback.events/ with the latest commit (b76de810aad37b2ee082492505473b17e2329bff) and still stack at spinning despite placing "LOCKS" in the `src/config/env.json`. 

```
~/.../kickback/app (pr/220)$yarn deploy:rinkeby
yarn run v1.15.2
$ yarn build:release:rinkeby && yarn now -f --local-config .deploy/now.rinkeby.json --public && yarn now alias --local-config .deploy/now.rinkeby.json && yarn now rm kickback-app-rinkeby --safe --yes
$ yarn setup --rinkeby && cross-env REACT_APP_ENV=rinkeby yarn build
$ scripts/setup.js --rinkeby
{
  "DEPLOYER_CONTRACT_ADDRESS": "0x291CD2C3e7Cf7627Fd8caC837635b7d00e9c9700",
  "API_URL": "https://rinkeby.api.kickback.events",
  "NUM_CONFIRMATIONS": 1,
  "ENV": "rinkeby",
  "DAI_CONTRACT_ADDRESS": "0xc4375b7de8af5a38a93548eb8453a498222c4ff2",
  "GIT_COMMIT": "b76de810aad37b2ee082492505473b17e2329bff",
  "LOGROCKET_TOKEN": "",
  "ROLLBAR_TOKEN": "e676d64e462b48d098a12db8a173598a",
  "BLOCKNATIVE_DAPPID": "27b3eac2-e46c-428a-9a0c-56cce2725d42",
  "MIXPANEL_ID": "28243587317e7b2d8a669dcce23302cb",
  "LOCKS": "https://staging-app.unlock-protocol.com/dashboard/"
}
```

<img width="1192" alt="Screenshot 2019-12-16 at 20 39 02" src="https://user-images.githubusercontent.com/2630/70941373-33f85880-2044-11ea-8efe-4692d6dfa903.png">

edsonayllon commented:
> @makoto Hi,I tried both local and https://rinkeby.kickback.events/ with the latest commit ([b76de81](https://github.com/wearekickback/app/commit/b76de810aad37b2ee082492505473b17e2329bff)) and still stack at spinning despite placing "LOCKS" in the `src/config/env.json`.
> 
> ```
> ~/.../kickback/app (pr/220)$yarn deploy:rinkeby
> yarn run v1.15.2
> $ yarn build:release:rinkeby && yarn now -f --local-config .deploy/now.rinkeby.json --public && yarn now alias --local-config .deploy/now.rinkeby.json && yarn now rm kickback-app-rinkeby --safe --yes
> $ yarn setup --rinkeby && cross-env REACT_APP_ENV=rinkeby yarn build
> $ scripts/setup.js --rinkeby
> {
>   "DEPLOYER_CONTRACT_ADDRESS": "0x291CD2C3e7Cf7627Fd8caC837635b7d00e9c9700",
>   "API_URL": "https://rinkeby.api.kickback.events",
>   "NUM_CONFIRMATIONS": 1,
>   "ENV": "rinkeby",
>   "DAI_CONTRACT_ADDRESS": "0xc4375b7de8af5a38a93548eb8453a498222c4ff2",
>   "GIT_COMMIT": "b76de810aad37b2ee082492505473b17e2329bff",
>   "LOGROCKET_TOKEN": "",
>   "ROLLBAR_TOKEN": "e676d64e462b48d098a12db8a173598a",
>   "BLOCKNATIVE_DAPPID": "27b3eac2-e46c-428a-9a0c-56cce2725d42",
>   "MIXPANEL_ID": "28243587317e7b2d8a669dcce23302cb",
>   "LOCKS": "https://staging-app.unlock-protocol.com/dashboard/"
> }
> ```

That locks config is wrong. Please follow the README. 

https://github.com/wearekickback/app/blob/b76de810aad37b2ee082492505473b17e2329bff/README.md#setup

edsonayllon commented:
@makoto I've updated `setup.js` for Rinkeby so the correct lock is automatically added. 2eee09a

makoto commented on src/routes/CreateEvent.js:
```diff
@@ -13,42 +15,170 @@ const CreateContainer = styled('div')`
+    let unlock = window.unlockProtocol.blockchainData()
+    let locks = Object.keys(unlock.locks).map(i => unlock.locks[i])
+    let bronze = locks.find(o => o.name.includes('Kickback Bronze'))
+    let gold = locks.find(o => o.name.includes('Kickback Gold'))
```
Is this part used to determine which lock I became the member of?

<img width="313" alt="Screenshot 2019-12-17 at 22 36 48" src="https://user-images.githubusercontent.com/2630/71040128-cc164080-211d-11ea-8454-cd7a1b09a8bb.png">

Does `unlock.locks` returns all the locks only purchased by the specific Ethereum account? I tested with 4 different metamask accounts where account 1~3 bought gold and account 4 bought bronze though it kept showing gold no matter which ETH account I switch into.

When I put `console.log({locks})` under account 4 it had both bronze and gold hence looks like it picked gold automatically.

I wonder if `unlock.locks` just returns all the available locks, rather than the lock purchased by the purchaser (=event organiser). cc @julien51

julien51 replied:
So, it is undocumented, but the way to get the locks for which the user has a key is to do this:

```
const data = window.unlockProtocol.blockchainData()
          if (data.keys) {
            // They should be indexed by lock.
            // Let's assume there is only one valid.
            const lockAddress = Object.keys(data.keys)[0]
          }
```

edsonayllon replied:
> @makoto I wonder if unlock.locks just returns all the available locks, rather than the lock purchased by the purchaser (=event organiser).

You're right. 

> @julien51 So, it is undocumented, but the way to get the locks for which the user has a key is to do this: [...]

`data.keys` returned all available locks as well for me. Tested by purchasing on a new account.

I used `data.transaction` instead to get the latest purchase. Then displayed the membership by the lock purchased. I left the full algorithm commented in `updateUnlockUser()`.

Should display the correct membership owned now. 

<img width="655" alt="Screen Shot 2019-12-17 at 7 16 11 PM" src="https://user-images.githubusercontent.com/34966228/71045320-84ce8680-2102-11ea-9a00-fb3484248c30.png">


c0eaed3

makoto commented:
LGTM @edsonayllon @julien51 . Thank you so much for being super patient with my constant feedback!

julien51 commented:
Let's merge this so we can release the bounty! ;) 
Happy new year everyone!

edsonayllon commented:
@makoto @jefflau Any updates on this?
